chronos/futures.nim

```diff
@@ -0,0 +1,369 @@
+#
+#                     Chronos
+#
+#  (c) Copyright 2015 Dominik Picheta
+#  (c) Copyright 2018-2023 Status Research & Development GmbH
+#
+#                Licensed under either of
+#    Apache License, version 2.0, (LICENSE-APACHEv2)
+#                MIT license (LICENSE-MIT)
+
+import std/[strutils, sequtils]
+import stew/base10
+import "."/[config, srcloc]
+export srcloc
+
+when chronosStackTrace:
+  when defined(nimHasStacktracesModule):
+    import system/stacktraces
+  else:
+    const
+      reraisedFromBegin = -10
+      reraisedFromEnd = -100
+
+when chronosStackTrace:
+  type StackTrace = string
+
+type
+  LocationKind* {.pure.} = enum
+    Create
+    Complete
+
+  CallbackFunc* = proc (arg: pointer) {.gcsafe, raises: [Defect].}
+
+  # TODO v4: consider removing this type and just use plain `CallbackFunc`
+  AsyncCallback* = object
+    function*: CallbackFunc
+    udata*: pointer
+
+  FutureState* {.pure.} = enum
+    Pending, Completed, Cancelled, Failed
+
+  FutureBase* = ref object of RootObj ## Untyped future.
+    when defined(chronosv4):
```

Comment:
Personally i do not like name `chronosv4`, but even if this name is going to be accepted it should be style compatible with all previous config settings so it should be `chronosV4`.

Reply:
renamed to `chronosPreviewV4` - once all of V4 is in place, the intent would be to remove the `when` along with the v3 support - this is just a way to allow the two versions to coexist temporarily during development - the alternative would be a branch